Documentation/DSLAPIWalkthough.md

```diff
@@ -0,0 +1,464 @@
+# Regex DSL API Walkthrough
+
+## Core
+
+### Core abstractions
+
+```swift
+public protocol RegexProtocol {
+  associatedtype Match
+  var regex: Regex<Match> { get }
+}
+
+public struct Regex<Match>: RegexProtocol {
+  public var regex: Regex<Match> { get }
+
+  public init<Content: RegexProtocol>(
+    _ content: Content
+  ) where Content.Match == Match
+   
+  public init<Content: RegexProtocol>(
+    @RegexBuilder _ content: () -> Content
+  ) where Content.Match == Match
+}
+```
+
+### Regex-y types
+
+```swift
+extension String: RegexProtocol {
+  public typealias Match = Substring
+  public var regex: Regex<Match>
+}
+
+extension String: RegexProtocol {
+  public typealias Match = Substring
+  public var regex: Regex<Match>
+}
+
+extension Character: RegexProtocol {
+  public typealias Match = Substring
+  public var regex: Regex<Match>
+}
+
+extension CharacterClass: RegexProtocol {
+  public typealias Match = Substring
+  public var regex: Regex<Match>
+}
```

Comment:
I think this will need its own section and probably some fleshing out. Is this meant to provide custom character classes, surface built-in ones, etc? Would we build as a scalar consumer with a derived protocol for character predicates, etc?